Improve and tests

john0isaac opened the pull request:
## Purpose

This PR sets adds unit tests for the python codebase with test coverage 84%. It also adds pedantic types for all api responses that follows the Microsoft Chat Protocol. It also fixes all of the typing errors, adds Mypy and improves the workflow for code quality analysis.

![image](https://github.com/Azure-Samples/rag-postgres-openai-python/assets/64026625/7a94116d-35c2-44b1-936f-1913666724fe)


## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Type of change

```
[ ] Bugfix
[ ] Feature
[x] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Code quality checklist

See [CONTRIBUTING.md](https://github.com/Azure-Samples/rag-postgres-openai-python/blob/main/CONTRIBUTING.md#submit-pr) for more details.

- [x] The current tests all pass (`python -m pytest`).
- [x] I added tests that prove my fix is effective or that my feature works
- [x] I ran `python -m pytest --cov` to verify 100% coverage of added lines
- [x] I ran `python -m mypy` to check for type errors
- [x] I either used the pre-commit hooks or ran `ruff` manually on my code.

john0isaac commented on .github/workflows/python-code-quality.yaml:
```diff
@@ -3,23 +3,34 @@ name: Python code quality
           run: |
-            python -m pip install --upgrade pip
-            pip install -r requirements-dev.txt
+            python3 -m pip install --upgrade pip
+            python3 -m pip install ruff
```
This is faster and works the same as a full installation.

john0isaac commented on pyproject.toml:
```diff
@@ -1,10 +1,20 @@
-[tool.ruff.lint]
-select = ["E", "F", "I", "UP"]
-ignore = ["D203"]
+[tool.mypy]
```
I ran mypy on the whole thing the only things left are the missing module stubs.
Don't know if you wanna do `#type: ignore` or add them to `pyproject.toml`

pamelafox replied:
Hm, I got a bunch more errors, let's tackle in a follow-up PR, mypy can be a bit annoying to satisfy.

john0isaac replied:
sure

john0isaac commented on requirements-dev.txt:
```diff
@@ -2,4 +2,8 @@
-pip-compile-cross-platform
\ No newline at end of file
+pip-compile-cross-platform
+pytest
+pytest-cov
+pytest-asyncio
+psycopg2-binary
```
Don't know why this is required for testing only not for the application itself. 🤔
It was erroring for `psycopg2` but I couldn't install it as it recommended installing `psycopg2-bianry`

john0isaac commented on src/fastapi_app/api_routes.py:
```diff
@@ -52,7 +54,7 @@ async def search_handler(query: str, top: int = 5, enable_vector_search: bool =
 
 
-@router.post("/chat")
+@router.post("/chat", response_model=RetrievalResponse)
```
This makes swagger API detect Schema for response and cast it according to this type.

john0isaac commented on src/fastapi_app/api_models.py:
```diff
@@ -17,3 +17,15 @@ class ThoughtStep(BaseModel):
     props: dict = {}
+
+
+class RAGContext(BaseModel):
+    data_points: dict[int, dict[str, Any]]
+    thoughts: list[ThoughtStep]
+    followup_questions: list[str] | None = None
+
+
+class RetrievalResponse(BaseModel):
+    message: Message
+    context: RAGContext
+    session_state: Any | None = None
```
This is the same as the frontend types to avoid dicts and automatic validation from pydantic.

john0isaac commented:
The test setup for the database still needs work.

john0isaac commented on tests/conftest.py:
```diff
@@ -0,0 +1,245 @@
+        yield test_client
+
+
+@pytest_asyncio.fixture(scope="function")
+async def db_session():
+    """Create a new database session with a rollback at the end of the test."""
+    engine = await create_postgres_engine_from_env()
+    async_sesion = async_sessionmaker(autocommit=False, autoflush=False, bind=engine)
+    session = async_sesion()
+    await session.begin()
+    yield session
+    await session.rollback()
+    await session.close()
+    await engine.dispose()
```
this is not used but it can be used for adding more tests in the future to validate the responses directly from the database.

pamelafox commented:
Thanks for the PR! Unfortunately been too busy with conferences and streams, but I will take a look as soon as I get a chance. I'm excited about all the improvements!

pamelafox commented:
@john0isaac 

I made a few typing improvements. Mypy seems to be confused by the editable app install, and I'm not sure how to make it happy.

There's also an issue with SQLAlchemy on these lines:

sql = text(fulltext_query).columns(id=Integer, rank=Integer)

I'm not sure what the issue is there. We could punt on making mypy happy, as I said in a comment, or we could spend a little more time trying to get the rest of the errors away, possibly via configuring what to ignore.

john0isaac commented:
Let me take a look and see what I can do.
I will let you know what I reach here.

pamelafox commented:
I asked about SQLAlchemy issue here: https://github.com/sqlalchemy/sqlalchemy/discussions/11600

pamelafox commented:
Asked pgvector for type annotation support here:
https://github.com/pgvector/pgvector-python/issues/82
I think we should ignore pgvector in pyproject.toml for now.

pamelafox commented:
The other issue is: "tests/conftest.py:225: error: Cannot instantiate abstract class "MockAzureCredential" with abstract attributes "__aexit__" and "close"  [abstract]"

I'm surprised to see that error as that's the same mock as azure-search-openai-demo, and I don't see the same error there.

john0isaac commented:
> The other issue is: "tests/conftest.py:225: error: Cannot instantiate abstract class "MockAzureCredential" with abstract attributes "**aexit**" and "close" [abstract]"
> 
> I'm surprised to see that error as that's the same mock as azure-search-openai-demo, and I don't see the same error there.

How can I reproduce this error?

Maybe it's this line that I changed? 
Original
<img width="717" alt="image" src="https://github.com/user-attachments/assets/f5e4448d-3e45-4972-a47b-bb0ca29e5a44">


What I changed

<img width="779" alt="image" src="https://github.com/user-attachments/assets/ad3c737b-b516-4271-beb4-eaabef496542">

john0isaac commented:
> I made a few typing improvements. Mypy seems to be confused by the editable app install, and I'm not sure how to make it happy.

fixed here https://github.com/Azure-Samples/rag-postgres-openai-python/pull/55/commits/addd3dafe6fadc4c299884f5930bf14c8d70b475

john0isaac commented:
> The other issue is: "tests/conftest.py:225: error: Cannot instantiate abstract class "MockAzureCredential" with abstract attributes "**aexit**" and "close" [abstract]"
> 
> I'm surprised to see that error as that's the same mock as azure-search-openai-demo, and I don't see the same error there.

AsyncTokenCredential is a Protocol
You just had to reimplement the functions in it's children without changing anything lol.

john0isaac commented:
> Asked pgvector for type annotation support here: [pgvector/pgvector-python#82](https://github.com/pgvector/pgvector-python/issues/82) I think we should ignore pgvector in pyproject.toml for now.

Done here https://github.com/Azure-Samples/rag-postgres-openai-python/pull/55/commits/39abb0fa8c7fede9987843eeacd9a1870bfe15c0

john0isaac commented:
> I asked about SQLAlchemy issue here: [sqlalchemy/sqlalchemy#11600](https://github.com/sqlalchemy/sqlalchemy/discussions/11600)

In the [sqlalchmey docs](https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.TextClause.columns), there are two ways of doing this:

## First Option / Passes mypy I tested it and pushed it here https://github.com/Azure-Samples/rag-postgres-openai-python/pull/55/commits/121b3417c509ff41f6aa893bf3bf348e40d5d4f9
```python
stmt = text("SELECT id, name, timestamp FROM some_table")
stmt = stmt.columns(
            column('id', Integer),
            column('name', Unicode),
            column('timestamp', DateTime)
        )

for id, name, timestamp in connection.execute(stmt):
    print(id, name, timestamp)
```

## Second Option / Original code - not working with mypy
```python
stmt = text("SELECT id, name, timestamp FROM some_table")
stmt = stmt.columns(
            id=Integer,
            name=Unicode,
            timestamp=DateTime
        )

for id, name, timestamp in connection.execute(stmt):
    print(id, name, timestamp)
```

john0isaac commented:
this https://github.com/Azure-Samples/rag-postgres-openai-python/pull/55/commits/5c17e9a404e772a40d24562c34aeb52f97441d13 fixes all of the errors related to mypy and the global_storage, it also follow the FastAPI way for creating application dependencies, the tests are faster now (I don't know why lol) maybe because I moved all of the code from the app startup 🤔

I created the routes folder to remove the noqa and to also follow the FastAPI app structure the rest of the files in the fastapi_app folder needs to be organized in folders too but I didn't want to make that change here too as it's not gonna affect anything.


I'm working on the last mypy error and will add mypy to the workflow run after I fix it.

john0isaac commented:
@pamelafox I did it 🎉 
All the errors are fixed and we are not using any `type: ignore`, tests are passing, app is working.
I also added mypy to the app_tests workflow as it requires a full installation of the packages (can't be added with ruff)

pamelafox commented on src/fastapi_app/dependencies.py:
```diff
@@ -0,0 +1,123 @@
+    """
+    Get the common parameters for the FastAPI app
+    """
+    load_dotenv(override=True)
```
We now have load_dotenv in multiple places, is that strictly necessary? I thought we'd only need it in the lifespan function, which would then call the rest?

john0isaac replied:
It is necessary as this is now not managed by the application lifespan, if you call it in the application lifespan you don't have to do it, if you call it anywhere else say in the scripts for setting up the database you need to refresh the env vars as they are not gonna be loaded there.

For the application to work it's not necessary to call it for other uses of this function you need to call it.

john0isaac replied:
The only solution would be for me to go everywhere this function is being called and add/check that load_dotenv is  called before it.

pamelafox replied:
Also- The scripts do also call load_dotenv in their if == "main" logic, is that enough? Are there other situations where we call the functions outside of the lifespan/scripts?

john0isaac replied:
ok, I will review that and see if that's the case then it's not neededz

john0isaac replied:
you were correct, It's was not needed resolved here https://github.com/Azure-Samples/rag-postgres-openai-python/pull/55/commits/da5220c9431c9c8ec637157f86fc0f18d4fe989e
you were always loading env in the main and create_app.

pamelafox commented on src/fastapi_app/dependencies.py:
```diff
@@ -0,0 +1,123 @@
+
+async def get_engine():
+    """Get the agent database engine"""
+    load_dotenv(override=True)
```
Same Q

john0isaac replied:
same response, this now no longer managed by the application lifespan. if you run it before running the app.

john0isaac replied:
resolved here https://github.com/Azure-Samples/rag-postgres-openai-python/pull/55/commits/da5220c9431c9c8ec637157f86fc0f18d4fe989e

pamelafox commented on src/fastapi_app/dependencies.py:
```diff
@@ -0,0 +1,123 @@
+
+async def get_openai_chat_client():
+    """Get the OpenAI chat client"""
+    azure_credentials = await get_azure_credentials()
```
It looks like this is now calling DefaultAzureCredential in multiple places. Generally we should try to create the credential once and then pass it around, as it can be slow to run. Is that possible?

john0isaac replied:
I intentionally did that authenticate each request separately but sure, I will look into making it called once.

john0isaac replied:
I did fix that one too here https://github.com/Azure-Samples/rag-postgres-openai-python/pull/55/commits/da5220c9431c9c8ec637157f86fc0f18d4fe989e

pamelafox commented on tests/mocks.py:
```diff
@@ -9,6 +10,17 @@ class MockAzureCredential(AsyncTokenCredential):
     async def get_token(self, uri):
         return MockToken("", 9999999999, "")
 
+    async def close(self) -> None:
```
I'm still perplexed as to why we didn't need this in azure-search-openai-demo, but maybe we have different mypy settings.

john0isaac replied:
I might be wrong but I think that sometimes when there are too much files/code Mypy doesn't show all of the errors 🤔 
I reviewed the setup there and it's the Same as ours except for the python version.

pamelafox commented:
@john0isaac Nice work making mypy happy, that's a tall feat. I have a few questions about possibly redundant calls.

pamelafox commented on src/fastapi_app/dependencies.py:
```diff
@@ -88,10 +86,11 @@ async def get_azure_credentials() -> azure.identity.DefaultAzureCredential | azu
         raise e
 
 
+azure_credentials = get_azure_credentials()
```
I usually try to avoid globally setting any variable immediately in a file, as it often makes testing harder (since mocks must be setup before that file gets imported). However, it appears tests are passing, so maybe it's okay in this case. We could move to a function updates a global if it ends up being problematic, I suppose.

john0isaac replied:
we are mocking the original module that's why it's no issue for any Default_azure_credentials

john0isaac replied:
I just understood why you did what you did with the life span thingy when I was tackling this, seems like a way more complicated issue than that. 😆

john0isaac commented:
There is only one issue left.. the engine is being created with each db_session which is being created with each request. It's not affecting the performance much but I wanted to raise it to you.
I will work on it tomorrow, maybe in another PR if you merged this one or here.

I tried solving this and I did it the same way you'd create an engine with FastAPI. (they just define it as a variable not inside a function in the dependencies folder) saying that our engine creation function is an async func which can't be used outside an event loop (tried workarounds for that using asyncio but nothing worked)
finally, I changed all of the function related to dealing with Postgres to become normal not async and everything worked.
But all of the tests failed 🤣 I tried to do a bunch of stuff for the tests to fix it but I came across lots of async errors and even when it worked there were some dangling background tasks for tests (tests were randomly passing and the tests that don't get a db session I think were failing)

pamelafox commented:
Ah, I see, thanks for raising that. It'd be best if we can reuse a session, like Flask-SQLAlchemy-Lite does: https://flask-sqlalchemy-lite.readthedocs.io/en/latest/session/
I do think it's less obvious how to architect app globals in FastAPI apps than in Quart/Flask apps, it's been interesting.

We're not in a rush to merge this, so I'll keep it open.

john0isaac commented:
session won't work, that was my first go to.
only accessible from `request.session.something` which is gonna be only available at the endpoint, I also looked at the globals for them and they have something called `app.state` which you can store stuff inside not the application level but it won't be initialized until the app get initialized which means that we can't use it in creating sessions.

After reconsidering everything I'm thinking of reusing the lifespan and globals approach for application wide global vars as it will work with the dependency I think. That is most likely what I will do if I didn't find a solution for the engine issue.
I will keep you posted here.

john0isaac commented:
Done ✅ 
I combined between the app.states and the dependencies and it looks amazing, everything is now created once in the lifespan then accessed from there. It's looking great! I'm satisfied with this now. 😄 
I also created the item data model that matches the db table (db table class is not usable as it's not a valid pydantic type) this now types the whole api.

<img width="1228" alt="image" src="https://github.com/user-attachments/assets/644e53af-7046-4e93-92af-999aa73640ad">

All of the endpoints has example responses now.

pamelafox commented on src/fastapi_app/dependencies.py:
```diff
@@ -0,0 +1,138 @@
+            await session.rollback()
+            raise
+        else:
+            await session.commit()
```
Just curious, where did you see this particular try/except/else pattern? Was this recommended in the SQLAlchemy docs? We're using Depends with SQLModel sessions in another repo as well, but we only have "yield session" there, wondering if we should change it.

pamelafox commented on src/fastapi_app/routes/api_routes.py:
```diff
@@ -0,0 +1,98 @@
+
+
+@router.get("/items/{id}", response_model=ItemPublic)
+async def item_handler(id: int, database_session: DBSession) -> ItemPublic:
```
Nit, it'd be nice to be consistent with database_session being either first or last in the list. It's different between these two first routes.

pamelafox replied:
Or is that because of path parameters? I am new to Depends

john0isaac replied:
nope I just missed it, resolved here https://github.com/Azure-Samples/rag-postgres-openai-python/pull/55/commits/17fc97f8493d49613a3a38d2747ef730234d21ad

pamelafox commented:
Cool! I just learnt about Depends. Seems like a nice pattern for FastAPI.

john0isaac commented on tests/mocks.py:
```diff
@@ -0,0 +1,22 @@
+MockToken = namedtuple("MockToken", ["token", "expires_on"])
+
+
+class MockAzureCredential(TokenCredential):
+    def get_token(self, uri):
+        return MockToken("", 9999999999)
+
```
did you notice this? see the test below
it errored as it's not supposed to be an AsyncTokenCredential

john0isaac commented on tests/test_dependencies.py:
```diff
@@ -0,0 +1,21 @@
+
+
+@pytest.mark.asyncio
+async def test_get_azure_credentials(mock_session_env, mock_default_azure_credential):
+    result = await get_azure_credentials()
+    token = result.get_token("https://vault.azure.net")
+    assert token.expires_on == 9999999999
+    assert token.token == ""
```
I used await here before get_token() and it was erroring "Not awaitable" with trace back to function definition in TokenCredential not AsyncTokenCredential.
